packages/core/src/browser/tree/tree-selection-state.ts

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { Tree } from './tree';
+import { DepthFirstTreeIterator } from './tree-iterator';
+import { TreeSelection, SelectableTreeNode } from './tree-selection';
+
+/**
+ * Class for representing and managing the selection state of a tree.
+ */
+export class TreeSelectionState {
```

Comment:
this class is instantiated outside of inversify, how one can rebind it?

Reply:
Well, OK, let's add a provider on top of that, but I would not inject an iterator or a dialog either. It smells similar to me.

Reply:
@akosyakov, could you please check what is wrong with the factory binding? Hopefully, you can easily spot the flaw. I was unable to figure it out. Here is the [commit](https://github.com/theia-ide/theia/pull/1438/commits/c9d6b1ad5aa7305429e9ddde520021c9018f3708). You can reproduce it by running the `tree-selection-state.spec.ts` from VSCode. Thanks!

Reply:
Ok, let’s go without it, if it is simpler

Reply:
OK, I have reverted it.